micro-rdk/src/common/data_store.rs

```diff
@@ -0,0 +1,512 @@
+                    ));
+                }
+                return Ok(Self {
+                    buffers,
+                    collector_keys,
+                });
+            }
+        }
+        Err(DataStoreError::DataStoreInitialized)
+    }
+
+    fn get_index_for_collector(
+        &self,
+        collector_key: &ResourceMethodKey,
+    ) -> Result<usize, DataStoreError> {
+        self.collector_keys
+            .iter()
+            .position(|key| key == collector_key)
+            .ok_or(DataStoreError::UnknownCollectorKey(collector_key.clone()))
+    }
+
+    fn get_maximum_message_size(&self) -> usize {
+        unsafe { DATA_STORE.len() / self.collector_keys.len() }
+    }
+}
+
+impl DataStore for StaticMemoryDataStore {
+    fn store_message(
+        &mut self,
+        collector_key: ResourceMethodKey,
+        message: SensorData,
+    ) -> Result<(), DataStoreError> {
+        let max_msg_size = self.get_maximum_message_size();
+        let buffer_index = self.get_index_for_collector(&collector_key)?;
+        let buffer = &mut self.buffers[buffer_index];
+        let encode_len = message.encoded_len();
+        if encode_len > max_msg_size {
+            return Err(DataStoreError::DataTooLarge);
+        }
+        if encode_len + 5 > buffer.vacant_len() {
+            return Err(DataStoreError::DataBufferFull(collector_key, message));
+        }
+        buffer
+            .push(0)
+            .map_err(|_| DataStoreError::DataWriteFailure)?;
+        let len_bytes = (encode_len as u32).to_be_bytes();
+        buffer.push_slice(&len_bytes);
+
+        let mut buf = BytesMut::with_capacity(message.encoded_len());
+        message.encode(&mut buf)?;
+        let mut buf_iter = buf.into_iter();
+        buffer.push_iter(&mut buf_iter);
+        Ok(())
+    }
+    fn read_next_message(
+        &mut self,
+        collector_key: ResourceMethodKey,
+    ) -> Result<StoreMessage, DataStoreError> {
+        let buffer_index = self.get_index_for_collector(&collector_key)?;
+        let buffer = &mut self.buffers[buffer_index];
+
+        if let Some(&&zero_byte) = buffer.iter().peekable().peek() {
+            if zero_byte != 0 {
+                return Err(DataStoreError::DataIntegrityError);
+            }
+            let _ = buffer.pop();
+        } else {
+            return Ok(StoreMessage::EndOfBuffer);
+        }
+        let mut encoded_len: [u8; 4] = [0; 4];
+        buffer.pop_slice(&mut encoded_len);
+        let encoded_len = u32::from_be_bytes(encoded_len) as usize;
+        let mut msg_vec: Vec<u8> = vec![0; encoded_len];
+        buffer.pop_slice(msg_vec.as_mut_slice());
+        let mut msg_bytes = BytesMut::with_capacity(encoded_len);
+        msg_bytes.put(Cursor::new(msg_vec));
+        Ok(StoreMessage::Message(msg_bytes))
+    }
+    fn clear(&mut self) {
+        for buffer in self.buffers.iter_mut() {
+            buffer.clear();
+            unsafe {
+                buffer.set_head(0);
```

Comment:
I guess there's no reasonable scenario now. Originally I thought I could make the data survive restart, so I figured it would be necessary to have a way to clear the store if, for example, there was a breaking proto change that would make uploading the existing data impossible.

Reply:
I think we should remove the method unless/until we need it.